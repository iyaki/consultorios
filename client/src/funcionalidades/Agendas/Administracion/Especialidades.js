import { Card, CardContent, CircularProgress, Grid, IconButton, TextField } from '@material-ui/core'
import DeleteIcon from '@material-ui/icons/Delete'
import SaveIcon from '@material-ui/icons/Save'
import SearchIcon from '@material-ui/icons/Search'
import ClearIcon from '@material-ui/icons/Clear'
import { useEffect, useRef, useState } from 'react'
import LeftFiltersGrid from '../../App/LeftFiltersGrid'
import Loading from '../../App/Loading'
import axios from 'axios'
import { useSnackbar } from 'notistack'
import { getErrorMessage } from '../../../utils/http/webApp'

export default function Especialidades () {
  const [especialidades, setEspecialidades] = useState([])
  const [loadingEspecialidades, setLoadingEspecialidades] = useState(true)
  const { enqueueSnackbar } = useSnackbar()
  const [especialidadFilter, setEspecialidadFilter] = useState('')

  const filteredEspecialidades = especialidades.filter((e) => e.nombre.toLowerCase().indexOf(especialidadFilter.toLowerCase()) >= 0)

  function fetchEspecialidades () {
    setLoadingEspecialidades(true)
    axios
      .get('http://localhost:8080/agendas/webapp/especialidades')
      .then(r => {
        if (r.status === 200) {
          setEspecialidades(r.data.data)
        }
      })
      .catch(r => {
        const message = getErrorMessage(r)
        enqueueSnackbar(message, { variant: 'error' })
      })
      .finally(() => {
        setLoadingEspecialidades(false)
      })
  }

  function createEspecialidad (nombre) {
    return new Promise((resolve, reject) => {
      axios
        .post(
          'http://localhost:8080/agendas/webapp/especialidades',
          {
            data: {
              nombre
            }
          }
        )
        .then(r => {
          fetchEspecialidades()
          resolve(r)
        })
        .catch(r => {
          const message = getErrorMessage(r)
          enqueueSnackbar(message, { variant: 'error' })
          reject(r)
        })
    })
  }

  function editEspecialidad (nombre, id) {
    return new Promise((resolve, reject) => {
      axios
        .patch(
          `http://localhost:8080/agendas/webapp/especialidades/${id}`,
          {
            data: {
              nombre
            }
          }
        )
        .then(r => {
          fetchEspecialidades()
          resolve(r)
        })
        .catch(r => {
          const message = getErrorMessage(r)
          enqueueSnackbar(message, { variant: 'error' })
          reject(r)
        })
    })
  }

  function deleteEspecialidad (id) {
    return new Promise((resolve, reject) => {
      axios
        .delete(`http://localhost:8080/agendas/webapp/especialidades/${id}`)
        .then(r => {
          fetchEspecialidades()
          resolve(r)
        })
        .catch(r => {
          const message = getErrorMessage(r)
          enqueueSnackbar(message, { variant: 'error' })
          reject(r)
        })
    })
  }

  // Este hook llama solo una vez a fetchEspecialidades debido al array vacio de dependencias que recibe como segundo argumento
  useEffect(fetchEspecialidades, [])

  return (
    <LeftFiltersGrid
      renderFiltros={<Filters filterCallback={setEspecialidadFilter} filterValue={especialidadFilter} />}
      renderData={
        <>
          <EspecialidadCard saveCallback={createEspecialidad} hideDelete />
          <EspecialiadesList
            loading={loadingEspecialidades}
            especialidades={filteredEspecialidades}
            editCallback={editEspecialidad}
            deleteCallback={deleteEspecialidad}
          />
        </>
      }
    />
  )
}

function EspecialiadesList ({ loading, especialidades, editCallback, deleteCallback }) {
  if (loading) {
    return (
      <Loading />
    )
  }

  return (
    <>
      {
        especialidades.map(e => (
          <EspecialidadCard
            key={e.id}
            especialidadNombre={e.nombre}
            especialidadId={e.id}
            saveCallback={editCallback}
            deleteCallback={deleteCallback}
          />
        ))
      }
    </>
  )
}

function EspecialidadCard ({
  saveCallback,
  especialidadId,
  especialidadNombre = '',
  hideDelete = false,
  deleteCallback
}) {
  const especialidadRef = useRef()
  const [saving, setSaving] = useState(false)
  const [deleting, setDeleting] = useState(false)

  function save () {
    if (saving) {
      return
    }

    setSaving(true)

    saveCallback(especialidadRef.current.value, especialidadId)
      .finally(() => {
        setSaving(false)
        if (!especialidadId) {
          especialidadRef.current.value = ''
        }
      })
  }

  function remove () {
    if (deleting) {
      return
    }

    setDeleting(true)

    deleteCallback(especialidadId)
      .finally(() => {
        setDeleting(false)
      })
  }

  return (
    <Card style={{ marginTop: '1%' }}>
      <CardContent>
        <Grid container spacing={1}>
          <Grid item xs={9}>
            <TextField label='Nombre' defaultValue={especialidadNombre} style={{ width: '100%' }} inputRef={especialidadRef} />
          </Grid>
          <Grid item xs={3} style={{ textAlign: 'center' }}>
            <IconButton onClick={save}>
              {
                saving
                  ? (<CircularProgress />)
                  : (<SaveIcon />)
              }
            </IconButton>
            {
              hideDelete || (
                deleting
                  ? (<CircularProgress />)
                  : (
                    <IconButton onClick={remove}>
                      <DeleteIcon />
                    </IconButton>
                    )
              )
            }
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  )
}

function Filters ({ filterValue, filterCallback }) {
  const inputRef = useRef()

  function clear () {
    inputRef.current.value = ''
    filterCallback('')
  }

  return (
    <Card style={{ height: '96%', margin: '2%' }}>
      <CardContent>
        <TextField label='Nombre' style={{ width: '65%' }} inputRef={inputRef} />
        <IconButton defaultValue={filterValue} onClick={() => { filterCallback(inputRef.current.value) }}>
          <SearchIcon />
        </IconButton>
        <IconButton onClick={clear}>
          <ClearIcon />
        </IconButton>
      </CardContent>
    </Card>
  )
}
