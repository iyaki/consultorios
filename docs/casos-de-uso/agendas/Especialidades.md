Dentro del sistema especialidad se refiere a una especialidad médica.

# Índice
- [Índice](#índice)
- [Casos de uso](#casos-de-uso)
  - [Registro de especialidades](#registro-de-especialidades)
    - [Dependencias](#dependencias)
    - [Precondiciones](#precondiciones)
    - [Descripción](#descripción)
    - [Excepciones](#excepciones)
    - [Comentarios](#comentarios)
  - [Edición de registros de especialidades](#edición-de-registros-de-especialidades)
    - [Dependencias](#dependencias-1)
    - [Precondiciones](#precondiciones-1)
    - [Descripción](#descripción-1)
    - [Excepciones](#excepciones-1)
    - [Comentarios](#comentarios-1)
  - [Eliminación de registros de especialidades](#eliminación-de-registros-de-especialidades)
    - [Dependencias](#dependencias-2)
    - [Precondiciones](#precondiciones-2)
    - [Descripción](#descripción-2)
    - [Excepciones](#excepciones-2)
    - [Comentarios](#comentarios-2)

# Casos de uso

## Registro de especialidades

### Dependencias

- Ninguna.

### Precondiciones

- Ninguna.

### Descripción

Los usuarios deben poder registrar especialidades médicas en el sistema
brindando un nombre para la misma con el fin de, posteriormente, poder
configurar agendas médicas.

### Excepciones

- No se debe permitir el registro de especialidades con un nombre vacío.
- No se debe permitir el registro de especialidades con un nombre compuesto
integramente de caracteres de espaciado (espacios o tabs).
- No se debe permitir el registro de especialidades con nombres repetidos.

### Comentarios

- Ninguno.

## Edición de registros de especialidades

### Dependencias

- [Registro de especialidades](#registro-de-especialidades).

### Precondiciones

- Contar con registros de especialidades.

### Descripción

Los usuarios deben poder editar el nombre de los registros de especialidades
médicas en el sistema brindando un nuevo nombre.

### Excepciones

- No se debe permitir el registro de especialidades con un nombre vacío.
- No se debe permitir el registro de especialidades con un nombre compuesto
integramente de caracteres de espaciado (espacios o tabs).
- No se debe permitir el registro de especialidades con nombres repetidos.

### Comentarios

- Ninguno.

## Eliminación de registros de especialidades

### Dependencias

- [Registro de especialidades](#registro-de-especialidades).

### Precondiciones

- Contar con registros de especialidades.

### Descripción

Los usuarios deben poder eliminar registros de especialidades médicas.

### Excepciones

- Ninguna.

### Comentarios

- Ninguno.
