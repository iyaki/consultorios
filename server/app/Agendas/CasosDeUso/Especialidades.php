<?php

declare(strict_types=1);

namespace Consultorio\Agendas\CasosDeUso;

use Consultorio\Agendas\Dominio\Especialidad;
use Consultorio\Agendas\Dominio\EspecialidadRepositoryInterface;
use Consultorio\Core\Aplicacion\UnitOfWorkInterface;

final class Especialidades
{
    public function __construct(
        private UnitOfWorkInterface $unitOfWork,
        private EspecialidadRepositoryInterface $especialidadRepository
    ) {
    }

    public function crear(EspecialidadDTO $especialidadDTO): EspecialidadDTO
    {
        try {
            $this->unitOfWork->beginTransaction();

            $nombre = $especialidadDTO->nombre();

            $this->assertEsNombreUnico($nombre);

            $especialidadId = $this->especialidadRepository->crearId();
            $especialidad = new Especialidad($especialidadId, $nombre);
            $this->especialidadRepository->add($especialidad);

            $this->unitOfWork->commit();

            return EspecialidadDTO::fromEntity($especialidad);
        } catch (\Throwable $throwable) {
            $this->unitOfWork->rollback();
            throw $throwable;
        }
    }

    private function assertEsNombreUnico(string $nombre): void
    {
        $especialidades = $this->especialidadRepository->findBy([
            'nombre' => $nombre,
        ]);

        if (count($especialidades) > 0) {
            throw new \Exception('Ya existe una especialidad con el nombre: ' . $nombre);
        }
    }
}
