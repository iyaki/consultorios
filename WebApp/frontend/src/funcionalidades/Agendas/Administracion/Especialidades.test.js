import { SnackbarProvider } from 'notistack'
import renderer from 'react-test-renderer'
import Especialidades from './Especialidades'

test('Comportamiento por defecto', () => {
  const tree = renderer
    .create(<SnackbarProvider><Especialidades /></SnackbarProvider>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
