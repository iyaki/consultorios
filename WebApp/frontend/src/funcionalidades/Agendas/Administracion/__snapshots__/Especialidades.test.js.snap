// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comportamiento por defecto 1`] = `
<div
  className="MuiGrid-root MuiGrid-container"
  style={
    Object {
      "height": "768px",
      "overflowY": "hidden",
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    style={
      Object {
        "height": "100%",
        "overflowY": "auto",
      }
    }
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "height": "96%",
          "margin": "2%",
        }
      }
    >
      <div
        className="MuiCardContent-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
        >
          Agendas &gt; Configurar Especialidades
        </p>
        <hr />
        <br />
        <div
          className="MuiFormControl-root MuiTextField-root"
          style={
            Object {
              "width": "65%",
            }
          }
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Nombre
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          defaultValue=""
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
    style={
      Object {
        "height": "100%",
        "overflowY": "auto",
      }
    }
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
        style={
          Object {
            "marginTop": "1%",
          }
        }
      >
        <div
          className="MuiCardContent-root"
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
            >
              <div
                className="MuiFormControl-root MuiTextField-root"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink={false}
                >
                  Nombre
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiInput-input"
                    defaultValue=""
                    disabled={false}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                  />
                </div>
              </div>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": "768px",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <div
          className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style={
            Object {
              "height": 40,
              "width": 40,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
